### PR TITLE
Evaluate ASR metrics for noisy and non-noisy subsets

### PR DESCRIPTION
**Pull request by janaab11**
Implements this functionality through:

- [x] an alternate entry point via CLI - `eevee asr --noisy`
- [x] a cleaning method to separate out the info tags from every utterance
- [ ] ~optionally, a yaml to define the noisy and non-noisy subsets~
- [x] add documentation for usage

### CODE REVIEW

File: eevee/metrics/asr.py [hunk 1]
```diff
@@ -649,3 +653,43 @@ def asr_report(
...
+        return "noisy"
+
+
+def process_noise_info(
+        true_labels: pd.DataFrame, pred_labels: pd.DataFrame
+) -> Tuple[Dict[str, pd.DataFrame], Dict[str, pd.DataFrame]]:
+
+    df = pd.merge(true_labels, pred_labels, on="id", how="inner")
+    
+    ## separate out info tags in transcripts and clean the original transcriptions
```
Comment (janaab11) [inline eevee/metrics/asr.py:682]: @sachinkum can you add this code ?

**Comment by janaab11** [comment]
tests failing is intentional. when the missing code is added, tests should/will not fail

File: eevee/metrics/asr.py [hunk 2]
```diff
@@ -649,3 +653,47 @@ def asr_report(
 
     else:
         return report
+
+
+def extract_info_tags(transcription: str) -> str:
```
Comment (janaab11) [inline eevee/metrics/asr.py:658]: missing

File: eevee/metrics/asr.py [hunk 3]
```diff
@@ -649,3 +653,47 @@ def asr_report(
...
+    return ""
+
+
+def remove_info_tags(transcription: str) -> str:
```
Comment (janaab11) [inline eevee/metrics/asr.py:662]: missing

File: eevee/metrics/asr.py (hunk 1 above)
Comment (sachinkum) [reply eevee/metrics/asr.py:698]: Done

File: eevee/metrics/asr.py (hunk 2 above)
Comment (sachinkum) [reply eevee/metrics/asr.py:658]: Added

File: eevee/metrics/asr.py (hunk 3 above)
Comment (sachinkum) [reply eevee/metrics/asr.py:669]: Added

**Comment by janaab11** [comment]
Ready to review and merge

File: tests/asr/test_process_noise_info.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from eevee.metrics.asr import extract_info_tags, remove_info_tags, define_noisy
+
+
+def test_process_noise_info():
+
+	clean_transcription = "This is an example"
+	noise_tags = ["", "<audio_silent>", "<inauidble>", "<background_speech>"]
+	noise_labels = [0, 0, 1, 1]
+
+	noisy_transcriptions = ["{}{}".format(clean_transcription, tag) for tag in noise_tags]
```
Comment (greed2411) [inline tests/asr/test_process_noise_info.py:10]: you sure you don't want spaces `{}{}` vs `{} {}` to be part of the tagged transcriptions and test it as well?
im not sure how transcriptions get tagged/predicted with or without spaces.

File: docs/metrics/asr.md [hunk 5]
```diff
@@ -76,6 +76,19 @@ This will add two csv files
...
+use the "--noisy" flag like:
+
+```shell
+eevee asr ./data/tagged.transcriptions.csv ./data/predicted.transcriptions.csv --noisy
+```
+
+Results are in two DataFrames - for each of `noisy` and `non-noisy` subsets, in order. An important note 
+here, is that the transcriptions in `tagged.transcriptions.csv` are expected to contain informational tags,
+like - `<audio_silent>`, `<inaudible>`, etc - which are normally removed when not using the "--noisy" flag.
+
```
Comment (greed2411) [inline docs/metrics/asr.md:91]: please add the python API way to use this in docs as well.

**Review by greed2411 — CHANGES_REQUESTED** [review]
@janaab11 @sachinkum have left comments on making them more readable (to me at the very least). I believe these changes requested would help us maintain it better in the long run.

please feel free to reply and make suggestions as well. 
and as always thanks for the PR.

File: tests/asr/test_process_noise_info.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from eevee.metrics.asr import extract_info_tags, remove_info_tags, define_noisy
+
+
+def test_process_noise_info():
+
+	clean_transcription = "This is an example"
+	noise_tags = ["", "<audio_silent>", "<inauidble>", "<background_speech>"]
```
[the remark this answers is not in the document]
Comment (janaab11) [reply tests/asr/test_process_noise_info.py:7]: yes. actually the contents of the info tag (i.e the text within the `<>`) dont really matter, beyond some keywords. they are added by annotators manually - so typos can be expected

File: tests/asr/test_process_noise_info.py (hunk 4 above)
Comment (janaab11) [reply tests/asr/test_process_noise_info.py:10]: i have seen both spaces and no spaces in the (ops) annotations - between info tags and words

counter-question: do spaces matter when calculating WER ? do you strip away spaces or consider them as words/characters ?

File: eevee/metrics/asr.py [hunk 7]
```diff
@@ -649,3 +653,69 @@ def asr_report(
...
+    transcription = re.sub(
+        "(\<.*?\>)|[`.,:;\[\]]]+|(<[a-z]+)|([a-z]+>)|([a-z]+_[a-z]+)|([a-z]+_)|(_[a-z]+)|~",
+        "",
+        transcription.strip(),
+    )
+    return transcription
+
+
+## change this if you want to change the definition of noisy
+def define_noisy(tags: List[str]) -> int:
```
[the remark this answers is not in the document]
Comment (janaab11) [reply eevee/metrics/asr.py:679]: do you mean examples of usage ? covered that in tests

File: tests/asr/test_process_noise_info.py (hunk 4 above)
Comment (janaab11) [reply tests/asr/test_process_noise_info.py:10]: either way, this is not an issue. `remove_info_tags()` strips trailing and leading spaces [here](https://github.com/skit-ai/eevee/blob/eea47e8973226e258aa4a14b44146243c2ce8fce/eevee/metrics/asr.py#L673)

File: eevee/metrics/asr.py [hunk 8]
```diff
@@ -1,10 +1,15 @@
...
 import itertools
 import json
 from collections import Counter
 from functools import reduce
 from operator import mul
 from typing import Any, Dict, List, Mapping, Tuple, Union
 
+from pandas import Series, DataFrame
+from pandas.core.arrays import ExtensionArray
+from pandas.core.generic import NDFrame
```
[the remark this answers is not in the document]
Comment (janaab11) [reply eevee/metrics/asr.py]: done

File: eevee/metrics/asr.py [hunk 9]
```diff
@@ -649,3 +653,69 @@ def asr_report(
...
+    if len(tags) > 0:
+        for tag in tags:
+            if "silent" in tag:
+                return 0
+        return 1
+    else:
+        return 0
+
+
+def process_noise_info(
```
[the remark this answers is not in the document]
Comment (janaab11) [reply eevee/metrics/asr.py:735]: done

File: eevee/metrics/asr.py [hunk 10]
```diff
@@ -649,3 +653,69 @@ def asr_report(
 
     else:
         return report
+
+
+def extract_info_tags(transcription: str) -> List[str]:
```
[the remark this answers is not in the document]
Comment (janaab11) [reply eevee/metrics/asr.py:654]: done

File: eevee/metrics/asr.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (janaab11) [reply eevee/metrics/asr.py]: done. added examples in the docstring

[the remark this answers is not in the document]
Comment (janaab11) [reply eevee/metrics/asr.py]: done

File: eevee/cli.py [hunk 11]
```diff
@@ -132,21 +136,56 @@ def main():
...
-                f'{args["<pred-labels>"].replace(".csv", "")}-dump.csv', index=False
-            )
-            ops.to_csv(
-                f'{args["<pred-labels>"].replace(".csv", "")}-ops.csv', index=False
-            )
-        else:
-            output = asr_report(true_labels, pred_labels)
+        if args["--noisy"]:
+
+            noisy_dict, not_noisy_dict = process_noise_info(true_labels, pred_labels)
```
[the remark this answers is not in the document]
Comment (janaab11) [reply eevee/cli.py:141]: added

File: docs/metrics/asr.md (hunk 5 above)
Comment (janaab11) [reply docs/metrics/asr.md:91]: added

**Comment by janaab11** [comment]
@greed2411 made changes. please check and merge whenever

File: tests/asr/test_process_noise_info.py (hunk 4 above)
Comment (greed2411) [reply tests/asr/test_process_noise_info.py:10]: @sachinkum can answer your question, on the best/correct way to do it with/without spaces.

**Review by greed2411 — APPROVED** [review]

**Comment by greed2411** [comment]
looks good to me, thanks for the PR @sachinkum & @janaab11.
great work! it means a lot.

**Pull request merged**